Bring intial changes into master

slashrsm opened the pull request.

slashrsm commented on src/hdprivatekey.js:
```diff
@@ -547,14 +547,6 @@ HDPrivateKey.prototype.toString = function() {
- * Returns the console representation of this extended private key.
- * @return string
- */
-HDPrivateKey.prototype.inspect = function() {
```
Just curious what was the reason for this removal? Sure, console.log() in production code isn't the best thing ever, but this could stay I suppose?

ClemensLey replied:
I was getting a warning that inspect is deprecated
https://nodejs.org/api/deprecations.html#deprecations_dep0079_custom_inspection_function_on_objects_via_inspect

There would have been a more consistent way to fix but I was preparing for a demo and wanted a quick fix. We should think about what to do with the inspect functions that are in almost all classes

slashrsm replied:
Created followup: #11

slashrsm commented on src/transaction/input/publickey.js:
```diff
@@ -29,9 +29,9 @@ inherits(PublicKeyInput, Input);
+  $.checkState(this.output instanceof Output, 'Malformed output found when signing transaction');
   sigtype = sigtype || (Signature.SIGHASH_ALL |  Signature.SIGHASH_FORKID);
-  var publicKey = privateKey.toPublicKey();
+  var publicKey = privateKey.toPublicKey()
```
Was this change intended?

ClemensLey replied:
probably not

slashrsm commented on src/transaction/input/scripthash.js:
```diff
@@ -0,0 +1,148 @@
+  this.threshold = 1;
+  this.redeemScript = redeemScript;
+  $.checkState(Script.buildScriptHashOut(this.redeemScript).equals(this.output.script),
+    'Provided redeemScript doesn\'t hash to the provided output');
```
We [should use backslashes](https://github.com/airbnb/javascript#strings--escaping) to delimit the string and avoid escaping.

slashrsm commented on src/transaction/input/scripthash.js:
```diff
@@ -0,0 +1,148 @@
+  sigtype = sigtype || (Signature.SIGHASH_ALL |  Signature.SIGHASH_FORKID);
+
+  var self = this;
+  var results = [];
```
Above we converted a similar hunk into something a bit more functionally oriented (filter(), ...). Should we use the same pattern here?

ClemensLey replied:
I think so

slashrsm commented on src/transaction/input/scripthash.js:
```diff
@@ -0,0 +1,148 @@
+};
+
+ScriptHashInput.prototype.isValidSignature = function(transaction, signature) {
+  // FIXME: Refactor signature so this is not necessary
```
Follow-up?

slashrsm replied:
created: #13 12

slashrsm commented:
- Fixed cosmetic comments
- Created followups for test coverage, figuring out inspect and refactor of isValidSignature
- Squashed and rebased commit history
